openedx-unsupported/configuration
# Secures Jenkins Analytics using Github OAuth plugin

Configures the Jenkins Global Security using the [Github OAuth plugin](https://wiki.jenkins-ci.org/display/JENKINS/Github+OAuth+Plugin) for authentication and authorization.

Adds a private ssh key to the `jenkins` user to allow ansible to continue using the Jenkins CLI tool when github oauth is enabled.

**JIRA tickets**: Implements [OLIVE-34](https://openedx.atlassian.net/browse/OLIVE-34), [OC-1528](https://tasks.opencraft.com/browse/OC-1528)

**Discussions**: Approach discussed with Gabe, Kevin, and Jacek.

**Dependencies**: None

**Screenshots**:

Github OAuth access request screen:
![github oauth request access](https://cloud.githubusercontent.com/assets/7556571/14519679/8f055b44-025f-11e6-811e-c6c83a33d71b.png)

Jenkins homepage shown to github members of `JENKINS_ANALYTICS_GITHUB_OAUTH_ORGANIZATIONS`:
![github org access](https://cloud.githubusercontent.com/assets/7556571/14519685/9f9fed34-025f-11e6-820e-637f959e12da.png)

Jenkins homepage shown to github users in `JENKINS_ANALYTICS_GITHUB_OAUTH_ADMIN_USERS`:
![admin access](https://cloud.githubusercontent.com/assets/7556571/14519708/c127c4f4-025f-11e6-895c-54ec41e26fd5.png)

Jenkins homepage shown to github users not in `JENKINS_ANALYTICS_GITHUB_OAUTH_ORGANIZATIONS` or `JENKINS_ANALYTICS_GITHUB_OAUTH_ADMIN_USERS`:
![github access denied](https://cloud.githubusercontent.com/assets/7556571/14519730/d863d8a6-025f-11e6-8f8d-7cfdf8c0aa51.png)

**Sandbox URL**: http://52.20.136.133:8080

**Testing instructions**:

To test via vagrant:
1. Install edx/configuration requirements in a virtualenv:
   
   ```
   virtualenv venv
   . venv/bin/activate
   git clone https://git@github.com/open-craft/configuration.git
   cd configuration
   git checkout -b jill/olive-34-github-oauth origin/jill/olive-34-github-oauth
   make requirements
   ```
2. Create a YAML file containing the jenkins analytics configuration.
   - See [jenkins_analytics README](https://github.com/open-craft/configuration/blob/jill/olive-34-github-oauth/playbooks/roles/jenkins_analytics/README.md) for details.  Please report any missing or confusing information found.
   - Also see [github oauth ansible-jenkins-credentials.yml.zip](https://github.com/edx/configuration/files/220490/github.oauth.ansible-jenkins-credentials.yml.zip) for an example.
3. Verify that you have access to https://github.com/edx-ops/edx-jenkins-job-dsl/
4. Build the vagrant instance, using the YAML file created above:
   
   ```
   cd vagrant/base/analytics_jenkins
   export VAGRANT_JENKINS_LOCAL_VARS_FILE=/path/to/your.yml
   vagrant up
   ```

Expected results are
- Visiting http://192.168.33.15:8080 or http://localhost:8080 will redirect you to a GitHub authentication screen
- If you accept the authentication (and you are a member of the `JENKINS_ANALYTICS_AUTH_JOB_BUILDERS`), you'll be allowed to view, run, and modify the installed AnalyticsSeedJob and the other scheduled analytics jobs.
- If you're a member of the `JENKINS_ANALYTICS_AUTH_ADMINISTRATORS` list, you'll be allowed to view, run and modify the jobs, and also be allowed to [Manage Jenkins](http://localhost:8080/manage) and [Credentials](http://localhost:8080/credential-store/).
- If you are not in `JENKINS_ANALYTICS_AUTH_JOB_BUILDERS` and not in `JENKINS_ANALYTICS_AUTH_ADMINISTRATORS` list, you will see a 403 page, with the message "Access Denied: <your username> is missing the Overall/Read permission".

**Author notes and concerns**:
1. The original jenkins unix system user security configuration is still available, and can be enabled by setting `JENKINS_ANALYTICS_AUTH_REALM: unix`.
2. To logout of Jenkins, you have to logout of Github first, and then you can logout of Jenkins.  You can revoke the Jenkins application access, via [Github Settings > OAuth applications](https://github.com/settings/applications).

**Reviewers**
- [x] @jibsheet
